app/com/arpnetworking/metrics/portal/scheduling/JobRepository.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2018 Dropbox, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics.portal.scheduling;
+
+import models.internal.scheduling.Job;
+
+import java.time.Instant;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * A storage medium for {@link Job}s. Essentially a Map that mints unique keys for new Job values.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public interface JobRepository {
+
+    /**
+     * Open / connect to the repository. Must be called before any other methods.
+     */
+    void open();
+
+    /**
+     * Close the repository. Any further operations are illegal until the next open() call.
+     */
+    void close();
+
+    /**
+     * Create or update a {@link Job}.
+     *
+     * @param job The {@link Job} to create or update.
+     */
+    void addOrUpdateJob(Job job);
+
+    /**
+     * Retrieve a previously-stored Job.
+     *
+     * @param id The id assigned to the Job by a previous call to {@code add}.
+     * @return The Job stored with that key.
+     */
+    Optional<Job> getJob(UUID id);
+
+    /**
+     * Get the last time that a job with a given UUID was run.
+     *
+     * @param id The id assigned to the Job by a previous call to {@code add}.
+     * @return The last time that that job was executed.
+     * @throws NoSuchElementException if no job has the given UUID.
+     */
+    Optional<Instant> getLastRun(UUID id) throws NoSuchElementException;
```

Comment:
Using an `Instant` here instead of a `ZonedDateTime` because it's nonsensical to talk about "the time zone in which the job was executed." This will result in needing some extra glue-code when we're computing the next time a Job should be executed: I feel strongly that it's worth it to avoid a proliferation of time zones that might get out of sync with each other and result in subtle bugs that are easy to forget to test.

Reply:
That's reasonable. Although since the schedule produces times with locale, and the job executes on that, the last time it executed could be easily expressed in the same way. However, your interpretation is strictly more correct. This may impact Christian's storage to ensure that the last run storage round-trip is not lossy.

Reply:
I think this is the right way to go, I came to a similar conclusion yesterday when implementing this functionality in the data model. I think the time zone should be confined to the view and that layer should be where we deal with `ZonedDateTime`.